### PR TITLE
Make tracestate limit the responsibility of a platform

### PR DESCRIPTION
**Pull request by SergeyKanzhelev**
Proposed update for the `tracestate` limit. Basically make it a platform responsibility. See #239

### CODE REVIEW

**Review by AloisReitbauer — CHANGES_REQUESTED** [review]
Needs language polishing

**Review by bogdandrutu — APPROVED** [review]

**Comment by bogdandrutu** [comment]
Looks good the message overall. I am not checking for the language.

**Comment by danielkhan** [comment]
I am not sure about it and I am happy to discuss this but isn't that a lot of ambiguity and reasoning around our thought-process for a spec?

**Review by AloisReitbauer — CHANGES_REQUESTED** [review]
This is not what we defined in the last meeting. Following this approach one entry can have 512 bytes. I would define a maximum size per entry and total number of entries. 

We should make it clear that dropping tracestate will break other tracing systems.

**Comment by AloisReitbauer** [comment]
This was the last decision made during TPAC:

Decision: Size limit
* Every entry limit of 128 bytes (including key). If violated it’s ALLOWED to evict a violating entry.
* Max number of entries is 5. If that’s violated, it’s ALLOWED to evict entries that exceed the limit (but only at the end!)
* There won’t be an overall limit of 512 anymore.
* An implementer is does NOT HAVE TO enforce those rules

**Comment by yurishkuro** [comment]
> Every entry limit of 128 bytes (including key)

any written records of justification?

**Comment by AloisReitbauer** [comment]
* Typical tracestate entries are between 36 and 77 bytes (according to #109)! In 512 bytes, this would allow between (6X and 14X tracers).
* To allow some head room, let's assume a entry-length of ~128 bytes. (would allow 4X tracers)
* Assuming, the most common number of tracers is 2 (APM vendor + Cloud provider). There might be cases of multiple APM vendors + multiple Cloud providers, but cases of 3/4 tracing systems will be the exception.
* "Edge" devices would usually only have one tracestate entry (since they initiate a call)

File: spec/20-http_header_format.md [hunk 1]
```diff
@@ -383,23 +383,32 @@ Rather, the entry would be rewritten to only include the most recent position:
...
-
-If the `tracestate` value has more than 512 characters, the tracer CAN decide to
-forward the `tracestate`. When propagating `tracestate` with the excessive
-length - the assumption SHOULD be that the receiver will drop this header.
+3.2.2](https://tools.ietf.org/html/rfc7230#section-3.2.2). Platform or
+tracing system may propagate them as they came, combine into a single
+header or split into multiple headers differently following the RFC
+specification.
+
+Platforms and tracing vendors SHOULD propagate at lease 512 characters
```
Comment (discostu105) [inline spec/20-http_header_format.md]: at "least"

Comment (AloisReitbauer) [reply spec/20-http_header_format.md]: just copied over from the notes :-)

File: spec/21-http_header_format_rationale.md [hunk 2]
```diff
@@ -98,17 +90,92 @@ configuration - certain header mutations may still be required. An example may b
...
+like positioning request in multiple distributed tracing graphs and
+providing backward compatibility with older protocols. But `tracestate`
+is not limited to this use and will drive innovations and new scenarios
+going forward.
+
+Opaque nature of `tracestate` value raises many questions and biggest
+one is what guarantees we can request from protocol implementors on
+propagating and storing this field.
+
+On one hand - field value should be small so implementors can satisfy
```
Comment (discostu105) [inline spec/21-http_header_format_rationale.md]: "On the one hand, "

File: spec/21-http_header_format_rationale.md [hunk 3]
```diff
@@ -98,17 +90,92 @@ configuration - certain header mutations may still be required. An example may b
...
+the requirement to pass the value all the time. It is especially
+critical for various messaging systems where metadata like `tracestate`
+is not paid for by the customer and provided for free.
+
+On other hand - in order for the field to be useful there should be some
```
Comment (discostu105) [inline spec/21-http_header_format_rationale.md]: "On the other hand, "

**Comment by yurishkuro** [comment]
> * Typical tracestate entries are between 36 and 77 bytes

@wu-sheng is that true of SkyWalking? You were telling me that you may pass the caller (name+endpoint) in the context, which can easily be much longer strings.

**Comment by SergeyKanzhelev** [comment]
@AloisReitbauer I explained in rationale portion of this PR why any arbitrary limits are not helpful. **TL;DR;** any limit spec will define will be too big for somebody to propagate and too small to somebody to use. Also limits defined as hard number of characters will not encourage vendors to minimize amount of information to propagate via `tracestate` and to limit this information to essentials only. Cloud vendors in their turn has a good incentive of providing better monitoring experience for customers to increase the amount of data they will propagate.

I understand that the proposal you've listed was discussed during Leon meeting, but alternative proposal from @yurishkuro that is reflected in this PR was suggested even earlier. And this is what we discussed later during KubeCon Seattle.

More from rationale doc in this PR:

> #### 2. Limit based on number of entries

> This proposal was made to eliminate the problem of a "greedy" vendor utilizes the full length of a `tracestate` and ultimately blocking the interoperability.

> Problem with this proposal is that limits proposed for the individual entry were still quite high for cloud providers.

> Also even though seemingly this proposal makes interoperability better, abuse of using multiple entries by a single vendor is still unavoidable.


>  #### 3. Make max length the decision of the implementor

> Header should be small so providers can satisfy the requirement to pass the value all the time.	
> Options above "protected" vendors from implementors who will take optional nature of `tracestate` and will not propagate `tracestate`. However the reality is that for many platforms even larger `tracestate` fields propagation is not a big issue.


> 512 bytes looks like a reasonable compromise.	So there was a proposal that the spec will suggest the length limits, and define the algorithm of trimming `tracestate`. However the final decision of an actual implementation to the platform.

File: spec/20-http_header_format.md (hunk 1 above)
Comment (SergeyKanzhelev) [reply spec/20-http_header_format.md]: thx!

File: spec/21-http_header_format_rationale.md [hunk 4]
```diff
@@ -98,17 +90,103 @@ configuration - certain header mutations may still be required. An example may b
...
+### Total size limit
+
+The field `tracestate` opens up rich interoperability and extensibility
+scenarios for vendors. There are defined use cases for the `tracestate`
+like positioning request in multiple distributed tracing graphs and
+providing backward compatibility with older protocols. But `tracestate`
+is not limited to this use and will drive innovations and new scenarios
+going forward.
+
+Opaque nature of `tracestate` value raises many questions and biggest
```
Comment (mtwo) [inline spec/21-http_header_format_rationale.md]: Replace with:

The opaque nature of `tracestate` introduces some tension between guarantees to vendors around propagating vendor-specific data with traces vs. storage and propagation constraints

File: spec/21-http_header_format_rationale.md [hunk 5]
```diff
@@ -98,17 +90,103 @@ configuration - certain header mutations may still be required. An example may b
...
+tempting and hard to implement feature.
+
+Talking about abuse - one thing this working group is doing is working
+on user-defined context propagation as part of [Correlation
+Context](https://w3c.github.io/correlation-context/) spec. Correlation
+context will provide a relieve valve for people who want to have freedom
+propagating relatively big payload across the components of a
+distributed trace.
+
+Specification allows removing and adding as many `tracestate` entries as
```
Comment (mtwo) [inline spec/21-http_header_format_rationale.md]: Replace the first sentence with "This specification allows removing and adding as many `tracestate` entries as an implementation needs.

**Review by mtwo — APPROVED** [review]
Two grammar nits. Please fix before submitting, but I'm approving the changes.

**Comment by danielkhan** [comment]
I found a few grammar problems like that one here:
> Options above "protected" vendors from implementors who will take
optional nature of `tracestate` and will not propagate `tracestate`.
However the reality is that for many platforms even larger `tracestate`
fields propagation is not a concern.

At the very least, I suggest to rephrase it to

> The options above "protected" vendors from implementors that will take
the optional nature of `tracestate` and will not propagate it.
However, the reality is that for many platforms propagating an even larger `tracestate` field is not a concern.

Still, it could be further edited down for clarity and conciseness.

How do we want to deal with that? Go with it and do a review after? But then we will again end up re-hashing discussions we had already.

**Comment by SergeyKanzhelev** [comment]
@mwear, @victorNewRelic can you please review from NewRelic perspective?

@yurishkuro any more concerns?

File: spec/21-http_header_format_rationale.md [hunk 6]
```diff
@@ -98,17 +90,103 @@ configuration - certain header mutations may still be required. An example may b
...
+distributed trace.
+
+This specification allows removing and adding as many `tracestate`
+entries as an implementation needs. This freedom is required to satisfy
+privacy and interoperability concerns. Thus all suggestions
+specification is making regarding the size of `tracestate` field will
+only be a recommendation that would improve vendors interoperability and
+cannot be "enforced" in practice.
+
+Possible solutions:
```
Comment (yurishkuro) [inline spec/21-http_header_format_rationale.md]: Are these "considered and rejected"? The caption is confusing.

**Comment by mwear** [comment]
@SergeyKanzhelev: We are ok with the limits suggested here.

File: spec/21-http_header_format_rationale.md (hunk 6 above)
Comment (SergeyKanzhelev) [reply spec/21-http_header_format_rationale.md]: I addressed feedback. Also just a note, it's in rationale doc, not a spec

**Review by AloisReitbauer — APPROVED** [review]

File: spec/20-http_header_format.md [hunk 7]
```diff
@@ -383,23 +383,33 @@ Rather, the entry would be rewritten to only include the most recent position:
...
+
+There are systems where propagating of 512 characters of `tracestate`
+may be expensive. In this case the maximum size of propagated
+`tracestate` header SHOULD be documented and explained. Cost of
+propagating `tracestate` SHOULD be weighted against the value of
+monitoring scenarios enabled for the end users.
+
+In situation when `tracestate` needs to be truncated due to size
+limitations platform of tracing vendor MUST truncate whole entries.
+Entries larger than `128` characters long SHOULD be removed first. Than
```
Comment (devinsba) [inline spec/20-http_header_format.md]:
```suggestion
Entries larger than `128` characters long SHOULD be removed first. Then
```

**Comment by plehegar** [comment]
(waiting on #260 for merge)

File: spec/21-http_header_format_rationale.md [hunk 8]
```diff
@@ -98,17 +90,109 @@ configuration - certain header mutations may still be required. An example may b
...
+distributed trace.
+
+This specification allows removing and adding as many `tracestate`
+entries as an implementation needs. This freedom is required to satisfy
+privacy and interoperability concerns. Thus all suggestions
+specification is making regarding the size of `tracestate` field will
+only be a recommendation that would improve vendors interoperability and
+cannot be "enforced" in practice.
+
+There solutions to address this problem were discussed:
```
Comment (nikmd23) [inline spec/21-http_header_format_rationale.md]:
```suggestion
Three solutions to address this problem were discussed:
```

File: spec/21-http_header_format_rationale.md [hunk 9]
```diff
@@ -98,17 +90,109 @@ configuration - certain header mutations may still be required. An example may b
...
+#### 1. Declare the arbitrary max length
+
+This is simplest and easy to understand solution. Proposals for the
+limit varied from `20` (size of a single `parent-id` with small
+identifier) to `1024` to fit up to 5 vendors with large tracestate
+entries. `512` is a nice median length.
+
+Extremely small proposal like `64` wasn't received well by vendors,
+while larger limits was unacceptable/undesirable for cloud vendors as
+COGS of implementing it are quite high.
```
Comment (nikmd23) [inline spec/21-http_header_format_rationale.md]:
```suggestion
costs of implementing it are quite high.
```

File: spec/20-http_header_format.md [hunk 10]
```diff
@@ -383,23 +383,33 @@ Rather, the entry would be rewritten to only include the most recent position:
...
+characters.
+
+There are systems where propagating of 512 characters of `tracestate`
+may be expensive. In this case the maximum size of propagated
+`tracestate` header SHOULD be documented and explained. Cost of
+propagating `tracestate` SHOULD be weighted against the value of
+monitoring scenarios enabled for the end users.
+
+In situation when `tracestate` needs to be truncated due to size
+limitations platform of tracing vendor MUST truncate whole entries.
```
Comment (nikmd23) [inline spec/20-http_header_format.md]: Pedantic: added missing comma

```suggestion
limitations, platform of tracing vendor MUST truncate whole entries.
```

File: spec/20-http_header_format.md [hunk 11]
```diff
@@ -383,23 +383,33 @@ Rather, the entry would be rewritten to only include the most recent position:
...
+Entries larger than `128` characters long SHOULD be removed first. Then
+entries SHOULD be removed starting from the end of `tracestate`. Note,
+other truncation strategies like whitelist entries, blacklist entries or
```
Comment (nikmd23) [inline spec/20-http_header_format.md]: whitelist and blacklist [can be problematic terms](https://github.com/graphite-project/graphite-web/issues/1569). In general, we should try stay away from them. 

```suggestion
other truncation strategies like safe list entries, blocked list entries or
```

**Review by nikmd23 — CHANGES_REQUESTED** [review]
I generally approve, but recommend accepting my change about whitelist/blacklist terminology.

**Comment by SergeyKanzhelev** [comment]
@nikmd23 thanks for review

**Review by nikmd23 — APPROVED** [review]
